controller.py

```diff
@@ -394,6 +394,70 @@ def apply_borrowing_policy(self, patron, license_pool):
             )        
         return None
 
+
+class AdminController(CirculationManagerController):
+
+    @property
+    def google(self):
+        return GoogleAuthService.from_environment(
+            self.url_for('google_auth_callback'), test_mode=self.manager.testing
+        )
+
+    def authenticated_admin_from_request(self):
+        """Returns an authenticated admin or begins the Google OAuth flow"""
+
+        access_token = flask.session.get("admin_access_token")
+        if access_token:
+            admin = get_one(self._db, Admin, access_token=access_token)
+            if admin and self.google.active_credentials(admin):
+                return admin
+        redirect_url = flask.request.args.get("redirect")
+        return redirect(self.google.auth_uri(redirect_url), Response=Response)
+
+    def authenticated_admin(self, admin_details):
+        """Creates or updates an admin with the given details"""
+
+        admin, ignore = get_one_or_create(
+            self._db, Admin, email=admin_details['email']
+        )
+        admin.update_credentials(
+            self._db, admin_details['access_token'], admin_details['credentials']
+        )
+        return admin
+
+    def signin(self):
+        """Redirects admin if they're signed in."""
+        admin = self.authenticated_admin_from_request()
+
+        if admin:
+            return redirect(flask.request.args.get("redirect"), Response=Response)
+
+    def redirect_after_signin(self):
+        """Uses the Google OAuth client to determine admin details upon
+        callback. Barring error, redirects to the provided redirect url.."""
+
+        admin_details, redirect_url = self.google.callback(flask.request.args)
+        if isinstance(admin_details, ProblemDetail):
+            return ProblemDetail
+
+        if not self.staff_email(admin_details['email']):
+            return INVALID_ADMIN_CREDENTIALS
```

Comment:
I don't think this should return a problem detail. It gets returned where you'd normally be redirected, so the response will be displayed to the user.